### PR TITLE
Remove Shell-Specific Logic

### PR DESCRIPTION
**Pull request by ekcasey**
[Readable](https://github.com/buildpacks/rfcs/blob/no-bash/text/0000-remove-shell-processes.md)

### CODE REVIEW

**Review by sclevine — APPROVED** [review]

File: text/0000-remove-shell-specific-processes.md
Comment (natalieparellano) [inline text/0000-remove-shell-specific-processes.md:150]: > A platform that wishes to always provide that functionality could add a buildpack like this to every build.

I recall something similar came up in discussions around prepare phase & project.toml (using buildpacks to perform actions instead of the platform - e.g., setting build time env) and I wonder if maybe that idea deserves its own RFC. It seems like a sticking point in these conversations has been "who is responsible" for creating these generic buildpacks, how would they be supported & distributed, etc.

**Comment by ekcasey** [comment]
Not sure why the github-actions bot assigned three people ... ?

File: text/0000-remove-shell-processes.md
Comment (hone) [inline text/0000-remove-shell-processes.md:91]:
```suggestion
One upside to our previous execution strategy was that it enable users to include environment variable references in arguments that were later evaluated in the container. To preserve this feature we can instead adopt a variation on the Kubernetes strategy for [environment variables interpolation](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config). If a buildpack or user includes `${<env>}` in the `command` or `args` and `{env}` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.
```

File: text/0000-remove-shell-processes.md
Comment (hone) [inline text/0000-remove-shell-processes.md:112]:
```suggestion
command = ["dotnet", "my-app.dll", "--url", "http://0.0.0.0:${PORT}"] # the default value of PORT would need to be provided in a layer
```

File: text/0000-remove-shell-processes.md
Comment (hone) [inline text/0000-remove-shell-processes.md:117]:
```suggestion
* If the buildpack authors believed that `--url` should be overridable they could set move the last two arguments from `command` to `args`.
```

File: text/0000-remove-shell-processes.md
Comment (jabrown85) [inline text/0000-remove-shell-processes.md:302]:
```suggestion
We could use `$(<env>)` for environment variable replacements instead of `${<env>}`. Currently, Kubernetes only replaces `$(FOO)` if `FOO` matches the name of an environment variable defined in the `env` section of the PodSpec, otherwise `$(FOO)` is passed through literally. The would allow the launcher to pick up where kubernetes leaves off and evaluate any remaining references.
```

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:91]: Is there a risk that in doing this we might somehow expose secrets? For example if the process type is:

```
./bin/server --key $MY_SECRET
```

If we substitute `MY_SECRET` in the process could it end up in the logs or somewhere that it might not when Bash substitutes it? 

Are they known libraries/mechanisms we can use to make sure we're on the right-side of the law here?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:168]:
```suggestion
The first command in this example did not need to change to behave as expected with the new API. Previously it was necessary to use a shell process in order to evaluate `${WORLD}`. Now, the shell is unnescessary. However, if the user wants to use more features of Bash interpolation they will need to explicitly invoke a shell, as shown in the second command in this example.
```

Comment (jkutner) [reply text/0000-remove-shell-processes.md:168]: I'm not sure I can complete sense of what "more features of Bash interpolation" includes. In particular, does this include things like `;`, `&`, and `&&` in the command? Those are the most common patterns I can think of (aside from env vars). What about different methods of quoting (`"` versus `'`)?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:173]: One of the things I worry about in this example is how it might unexpectedly fail. I expect that most users will do the following:

1. upgrade
2. try the exact same command they used before (i.e. no `bash -c`)
3. watch it fail
4. realize they need to add `bash -c`

In step 3, is it possible that something very bad happens if their process kinda works but not as they expected?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:210]: I'd still need to move the script in the `exec.d` though right? are there any restrictions like filename or permissions that might change? (I'm sure these are all fine but would like to capture)

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:237]: Is this true even when the use `bash -c` in their command?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:309]: I'd like to bring this into scope

File: text/0000-remove-shell-processes.md
Comment (jkutner) [inline text/0000-remove-shell-processes.md:352]: I'm not sure this is winning example. There are 3 different impls of something that should (IMO) be a native feature. The fact that these haven't converged on a single Procfile buildpacks is worrisome and not a pattern I'd like to repeat.

File: text/0000-remove-shell-specific-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-specific-processes.md]: @natalieparellano I agree that an RFC for this sounds like a good idea.

If we, as a project, wanted to provide buildpacks to make it easy for platforms or buildpck ecosystems to adopts extensions or conventions, that would deserve its own RFC. However, I don't think we need to block this RFC on that. Whether or not we take the extra step of providing common buildpacks (and I suspect we should) an app->lifecycle API is fundamentally untenable and I would argue that we need to remove this feature either way.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:112]: I'll admit to knowing precious little about the `dotnet` CLI, but I copied this command directly from the buildpack. And it seems to agree with [this blog post](https://andrewlock.net/5-ways-to-set-the-urls-for-an-aspnetcore-app/#command-line-arguments) I just found.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:117]: https://github.com/buildpacks/rfcs/pull/168#discussion_r659030633

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:91]: I think we are right that there is some cause for concern here. I think we should change this RFC and switch the interpolation notation to something like `$((MY_SECRET))` so we never interpolate environment variables that are intended for Bash to interpolate in a command like `["bash", "-c", "./bin/server --key ${MY_SECRET}"]. It's a little sad that we need yet another notation but it seems good to remove any ambiguity with regards to what gets interpolated by k8s, vs the launcher, vs Bash (if it is explicitly included in a command).

In this example are you imagining that it is the user you provides `$MY_SECRET` dynamically at runtime? I would argue that this is not a safe thing to do with secrets because they end up in the container config. Secrets should be provided via a safe mechanism like bindings and buildpacks should be responsibility for protecting the secret, which hopefully will not require passing it directly on the command line (even as a reference for Bash to interpolate) in the vast majority of use cases.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:168]: To avoid this whole question I think we should switch from `${NAME}` to `$((NAME))` for our env var notation. Therefore nobody is tempted to try things like `${NAME:-default-name}`. We should keep it as simple as possible and avoid surprises.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:173]: I agree that they might upgrade and fail but I don't think there is any danger of the process even kinda working. In this case the launcher would be trying to directly invoke an executable called `'for opt in $JAVA_OPTS; do echo $opt; done'`

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:210]: Good point. We should capture two things:
1. The script must move to `<layer>/exec.d/`
2. The runtime user must have permission to execute this file (e.g. executable bit should be set).

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:237]: Yes I believe this is still true. `exec.d` helpers currently run for both direct and shell processes.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:315]: > Often, those modifications aren't even buildpack specific, but app specific

Not sure I follow why this is relevant.

What I am trying to say is that a single buildpack could continue to provide support for .profile scripts without forcing users to migrate to the exec.d interface, if that was desirable.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:309]: After thinking long and hard about it I am very opposed to bringing this into scope. I don't think contracts between the app source and the lifecycle are supportable. The platform API should be a contract between a platform and the lifecycle, to enable predictable behavior and a clean upgrade story. Otherwise the same app will behave different on different platforms depending on the platform API in use. I realize the user must have some awareness of the "launch" api to understand argument handling for the resulting image but at least they won't need to change their source code to run on different platforms. This RFC is trying hard to shrink the surface area of platform API, that the app developer needs to be aware of.

If we wanted to respect config like this in the app dir, a buildpack should provide that functionality. App developers can included the exact buildpack they want to get the exact outcome they want without having to keep track of the versions of the platform API currently implemented by the platforms they use (which is entirely outside their control).

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:352]: I love to see an RFC to provide utility buildpacks as part of the CNB project, that implement extension specs or common conventions.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:112]: Sorry, I think I found the source of confusion - https://github.com/buildpacks/rfcs/pull/168/commits/c97e7f22e989bb6aa45570d8f7da1ca86631ffa8

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:117]: https://github.com/buildpacks/rfcs/pull/168#discussion_r661799403

File: text/0000-remove-shell-processes.md
Comment (samj1912) [reply text/0000-remove-shell-processes.md]: `$(ENV)` should work? (the k8s convention). It only works in k8s for env vars directly defined  in the spec, the rest are left untouched.

File: text/0000-remove-shell-processes.md
Comment (samj1912) [inline text/0000-remove-shell-processes.md:166]: Could we support these use cases by default? GIven that most envsubst libs already support it - 

https://github.com/a8m/envsubst
https://github.com/drone/envsubst

This would also be useful as it would end up in the metadata and users could see default values for these variables as set by processes.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md]: Added in https://github.com/buildpacks/rfcs/pull/168/commits/b49861f13dd11b080099b0da7ac8b49e9611f113

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:166]: I _just_ decided to take this in a different direction - https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3

Comment (ekcasey) [reply text/0000-remove-shell-processes.md:166]: I found some of the comment from the original k8s implementation of this illuminating - https://github.com/kubernetes/kubernetes/pull/7936#issuecomment-100950516

Comment (samj1912) [reply text/0000-remove-shell-processes.md:166]: Was going through the same thing.


On a separate note - I guess it should be a trivial change in the above libs to add support for a different marker instead of `{`. WIll see if they are open to it. OTOH I guess we could always start simple and add it later if needed. I would've like a way to specify a default value at the very least though.

Comment (samj1912) [reply text/0000-remove-shell-processes.md:166]: FWIW - just needs a change at https://github.com/drone/envsubst/blob/7bf45dbf5372da60a12eab36b9c8a63c99b92ea5/parse/scan.go#L174

File: text/0000-remove-shell-processes.md
Comment (dmikusa-pivotal) [inline text/0000-remove-shell-processes.md:86]: I think there could be some nuance here that we need to document when switching from a shell process to direct regarding signals. Many platforms support the concept of a graceful shutdown, which is a SIGTERM followed by a pause, followed by SIGKILL on your process.

When you run as a shell process, pid 1 is your shell which generally traps and handles signals as you'd expect. When you switch to running your process as direct, it becomes pid 1 and it may or may not handle all the signals correctly (you can generally tell if it's not handling them correctly by running a container and trying to interrupt it with ctrl+c, if that works it's handling signals correctly). 

Anyway, I think we need to call out this consideration so that when buildpack authors are migrating their buildpacks they think about this nuance and ensure it's handled correctly.

File: text/0000-remove-shell-processes.md
Comment (hone) [reply text/0000-remove-shell-processes.md:359]: @ekcasey isn't this already true by us removing support fro `<app>/.profile` since this changes the contract? I get you're trying to simplify/minimize the scope here, but bringing this into scope or not is still changing that contract.

File: text/0000-remove-shell-processes.md
Comment (jkutner) [reply text/0000-remove-shell-processes.md:402]: https://github.com/buildpacks/rfcs/pull/175

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md]: updated to use k8s style notation in https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:359]: While you are right that removing `<app>/.profile` modifies this contract, it shrinks it in a way that should reduce future occurrences of this problem. Adding `<app>/.exec` modified this contract without helping to mitigate this problem in the future.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:86]: After some reading I suspect the default signal handling behavior is actually better for a direct process than a shell process. Bash does not forward the SIGTERM to the child process by default and therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. Therefore a buildpack author that is not thinking about this explicitly might prevent the appropriate handling of a sigterm if they default to creating a shell process.

https://veithen.io/2014/11/16/sigterm-propagation.html

cc @hone, @dmikusa-pivotal

Comment (sclevine) [reply text/0000-remove-shell-processes.md:86]: > therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. 

Surprisingly, it's even worse than this. A simple bash invocation (e.g., `bash -c "tail -f /dev/null"`) is treated differently by bash than a complex one (e.g., `bash -c "true && tail -f /dev/null"`). In the former case, bash `exec`s itself away, resulting in a single process. So from the user's perspective, whether this works or not is quite arbitrary.

Comment (dmikusa-pivotal) [reply text/0000-remove-shell-processes.md:86]: Hmm, why isn't anything ever easy?  Lol.

The behavior I'd observed is with apps that don't handle signals at all. If you run them as PID1, you can't terminate the container. It's SIGKILL or the app keeps running. However, if you wrap that in `bash -c` it'll terminate as expected. I guess I didn't dig deep enough to check how Bash was in turn stopping that app, based on what's linked here it is just killing it as well and not propagating signals.

Regardless, there seems to be enough complexity here that I still think this warrants a mention in this RFC document. It's nuanced behavior that could easily be missed by a buildpack author, if we call attention to it, that gives authors a chance to investigate what it means to them. 

Even if it's as simple as "When running a process as direct, your process will always run as PID1. Buildpack authors, please be aware this may change the behavior of your application at shutdown and you should review accordingly.", I think that give's authors a heads up.

Comment (dmikusa-pivotal) [reply text/0000-remove-shell-processes.md:86]: One more "fun" point on this subject. If you use `sh -c` instead of `bash -c` you get more different behavior because at least on Ubuntu, `/bin/sh` is not `/bin/bash`, it's `/bin/dash`. I was running a process this morning with `sh -c` and it completely ignored SIGTERM.

Comment (ekcasey) [reply text/0000-remove-shell-processes.md:86]: @sclevine Are you aware of any documentation regarding when bash does or doesn't exec the child process. When I tried out you example commands I observed Bash replacing itself in memory, even for the more complex command:
```
docker run --rm --name test --entrypoint bash ubuntu:bionic -c 'true && tail -f /dev/null'
```
```
docker exec test ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
root           1       0  0 14:41 ?        00:00:00 tail -f /dev/null
root           7       0  0 14:41 ?        00:00:00 ps -ef
```

Comment (ekcasey) [reply text/0000-remove-shell-processes.md:86]: Okay, if I am understanding correctly, I think there are relevant two cases here:
1. the command is a complicated Bash invocation
2. the command is a simple Bash invocation

My understanding of the behavior (based on @sclevine 's description and my experimental verification):

**case 1**: A complicated command cannot be executed directly (`true && tail -f /dev/null` is not a valid direct process) and therefore `bash -c` must be explicitly included in the process. Therefore, wrt signal handling, **nothing changes**.
**case 2**: Bash is already `exec`ing itself away leaving the process as PID1. If this is converted to a direct process the same process will be PID1. Therefore, wrt signal handling, **nothing changes**.

**Summary**: With regards to signal handling nothing changes. App developers and buildpacks authors do not need to worry about changes to signal handling.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:166]: @samj1912 can this comment be resolved? or do you still want to advocate for more env substitution functionality, including defaults?

Personally I think we should go with the simple version for now.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:112]: I'm gonna assume this can be resovled

File: text/0000-remove-shell-specific-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-specific-processes.md]: @natalieparellano can this comment be resolved given the discussion has moved to https://github.com/buildpacks/rfcs/pull/175? Or do you feel that this RFC need further language to address this before approval?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [reply text/0000-remove-shell-processes.md:274]: No, i mean if they use `bash -c` will `.profile` still be not be run? I feel like that might be unexpected.

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:86]: @dmikusa-pivotal & @hone do you feel that this comment can be resolved given the explanation above describing how this does not affect end users or buildpack authors?

File: text/0000-remove-shell-processes.md
Comment (jkutner) [reply text/0000-remove-shell-processes.md:359]: I've created a an RFC that proposes a new mechanism to replace the `.profile` behavior (thus moving the contract to be between a builder<>app). https://github.com/buildpacks/rfcs/pull/175

**Comment by jkutner** [comment]
I'm supportive of this proposal, but I'd like to see us figure out how we'll replace the gaps left behind by not including shell-specific logic in the lifecycle. I've started a proposal in https://github.com/buildpacks/rfcs/pull/175, which combined with a forthcoming proposal for "Default Buildpacks in a Builder" may accomplish this. But I'd like to get alignment on these proposals before finalizing this one (#168)

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:274]: Yes, `.profile` will still not be run even when the command includes `bash -c`. Part of the goal here is to avoid doing anything complicated or magic. In order to run `.profile`s we would need to change the provided command from something like `['bash', '-c', 'echo hi']` to `['bash', '-c', 'source <app>/.profile && echo hi']`. Modifying the provided command if and only if it starts with `bash -c` feel like too much instrusive magic to me.

File: text/0000-remove-shell-processes.md
Comment (dmikusa-pivotal) [reply text/0000-remove-shell-processes.md:86]: I think this discussion has been good around signals. It seems unlikely that it's going to cause issues, so I think we're good there.

Maybe one final suggestion with this sentence...

>If `bash` or `cmd.exe` is required it should be included in `command`.

I think "required" might be nebulous and assumes you know what is required. It's easy to overlook restrictions of running as direct/PID1. Perhaps we could have something like:

>If currently using `direct = false`, the recommended migration path for buildpack authors is to include `bash` or `cmd.exe`  in `command`. This will result in the same behavior as `direct = false`. Before switching to `direct = true`, it is recommended you confirm environment variable interpolation is not required and that your process can function as PID1.

I guess it doesn't necessarily have to be in this line, but I think we should offer some migration advice and list the concerns an author should review before switching to `direct = true`.

**Review by nebhale — APPROVED** [review]

**Review by samj1912 — APPROVED** [review]

**Comment by ekcasey** [comment]
@jkutner 
> I'm supportive of this proposal, but I'd like to see us figure out how we'll replace the gaps left behind by not including shell-specific logic in the lifecycle. I've started a proposal in #175, which combined with a forthcoming proposal for "Default Buildpacks in a Builder" may accomplish this. But I'd like to get alignment on those proposals before finalizing this one (#168)

I added language to this RFC, making its implementation contingent on a .profile buildpack. Please see https://github.com/buildpacks/rfcs/pull/168/commits/f6cb6f4be2001c9f5fecd953e05fa36cc8bc19f0. I am hoping that with this caveat we can approve this RFC so we can start making progress while we figure out the nitty-gritty of the utility buildpacks. I assume that the spec and implemenntation of this RFC will take some time, while implementation of a `.profile` buildpack can proceed very quickly once we are done bikeshedding on #175.

**Review by jkutner — APPROVED** [review]

File: text/0000-remove-shell-processes.md
Comment (ekcasey) [reply text/0000-remove-shell-processes.md:87]: @dmikusa-pivotal please see https://github.com/buildpacks/rfcs/pull/168/commits/14066e50a3ee999fe0ff5249362e43c52169cac5

Comment (dmikusa-pivotal) [reply text/0000-remove-shell-processes.md:87]: Looks great, thanks!

**Comment by sclevine** [comment]
FCP ending 8/4

File: text/0000-remove-shell-processes.md
Comment (ryanmoran) [inline text/0000-remove-shell-processes.md:113]:
```suggestion
command = ["dotnet", "my-app.dll", "--urls", "http://0.0.0.0:$(PORT)"] # the default value of PORT would need to be provided in a layer
```

Should this follow the k8s interpolation format since this is no longer a shell process?

**Review by hone — APPROVED** [review]
@ekcasey Do we want to standardize "Depends On" for future RFCs?

File: text/0000-remove-shell-processes.md
Comment (hone) [reply text/0000-remove-shell-processes.md:113]: @ekcasey please take a look.

**Comment by ekcasey** [comment]
/queue-issue buildpacks/spec "Remove Shell-Specific Logic from Buildpack API" api/buildpack
/queue-issue buildpacks/spec "Update handling of user-provided command and args" api/buildpack
/queue-issue buildpacks/lifecycle "Remove Shell-Specific Logic"
/queue-issue buildpacks/pack "Update process types in inspect-image output to support command slice"
/queue-issue buildpacks/docs "Remove Shell-Specific Logic"

File: text/0000-remove-shell-processes.md
Comment (jabrown85) [inline text/0000-remove-shell-processes.md:359]: The pros/cons look funny in the PR markdown preview. I think you may need an additional line break between lists

**Review by jabrown85 — APPROVED** [review]

**Pull request merged**
